### DISCUSSION TITLE
Restrict access to all not whitelisted folders of a drive

### DISCUSSION THREAD

**Discussion by ghost** [sandboxie-plus/Sandboxie#2633] [Q&A Feedback]
Hello,

I would like to give sandboxed apps full access to root of A: drive and A:\abc folder and block access to all other folders on the same drive like A:\xyz etc.

In other words, is it possible to whitelist root of a drive and some folder on the same drive and block access to all other folders?

```
if((Test-Path A:) -or (Test-Path A:\abc)) {
Write-Host("Full access")
} else {
Write-Host("No access")
}
```

**Comment by bastik-1001** [comment]
I could not get access to the root access to function openly, but maybe someone else can find a better way to configure it.

In my test `UseRuleSpecificity=y` was required to be set. I tried to restrict access to D:\* and now programs like notepad (editor) can't access D:\ and D:\random\

I used the following entries for the box (normal box, yellow icon):

>UseRuleSpecificity=y
ClosedFilePath=D:\*
NormalFilePath=D:\demo\*
ClosedFilePath=D:\demo\na.txt
NormalFilePath=D:\test.txt

With notepad I could load D:\demo and see all its files (demo.txt and na.txt), it could load demo.txt located there, but due to the closed path for na.txt, the file could be seen by notepad, but it could not be opened nor was the content shown.  Due to D:\test.txt being a normal file path for Sandboxie again (last entry in the ini), I could load test.txt from the root folder, by directly opening it.

It should be possible to repeat this with `OpenFilePath` so those files can be fully accessed by the program. Edit: With the entries that contain NormalFilePath, being changed to OpenFilePath, it remains the same, those files can be accessed and when they are written to, it happens outside the sandbox. D:\demo\na.txt can't be read, written or deleted.

**Comment by ghost** [comment]
With the following settings I still can't save on root of a drive
```
OpenFilePath=A:\
ClosedFilePath=A:\*\*
```

**Reply by bastik-1001** [reply]
Maybe this is a limitation or a bug. I can't tell.

**Comment by DavidXanatos** [comment]
mmh... i would have thought that something like this:
```
WriteFilePath=E:\*\*
UseRuleSpecificity=y
OpenFilePath=E:\*
OpenFilePath=E:\Test\*
```

should do the trick, but seams I cant access the files on the root, not sure why, will have to debug this.

**Reply by ghost** [reply]
Thanks. It has been like that... for years

**Reply by DavidXanatos** [reply]
```
        //
        // if we have a pattern like C:\Windows\,
        // we still want it to match a path like C:\Windows,
        // hence we add a L'\\' to the path and check again
        //
```

this is what breaks this use case

´´´E:\Test.txt matches E:\*´´´
but
´´´E:\Test.txt\ matches E:\*\*´´´

I'll have to think about how to fix that....

I'm afraid that just removing this workaround will break other use cases or possible even some templates, I can check the templates, but the custom rules users created will be a problem.

unless we find a clever way to workaround the workaround

**Reply by DavidXanatos** [reply]
I have worked out an acceptable workaround, when we fall into the auxiliary match we set a flag and if there is later a match even if its less specific it will overrule previous bodged match.

to be noted here is that directory listing of all directories in the opened root will be allowed, as at this level the folders are just like the files which are intended to be accessible.
Howeever any access attempts to the items in those folders will result in an access denied.

This fix will not be included in in 1.7.2 as its not well tested yet, it will be included in a pre-release build after the next 1.7x based final build.

**Reply by ghost** [reply]
Thanks for 1.8.0 release. I am not sure how was I suppose to configure it to block access to `A:\*\` and allow full access to root of `A:\`

BTW: If you add `A:\` in resource access -> files, it changes to `A:\*`. In Sandboxie.ini, it is just `A:\`. I guess it is how it should be as per https://github.com/sandboxie-plus/Sandboxie/issues/2039#issuecomment-1193083437

**Reply by offhub** [reply]
#### For 1.8.0:

[Reference](https://www.wilderssecurity.com/threads/sandboxie-plus-v1-8-0.450696/#post-3134936)
```
UseRuleSpecificity=y
ClosedFilePath=A:\*\*
OpenFilePath=A:\*
OpenFilePath=A:\Test\*
```
> **Note**: The trailing * is required!

---

>BTW: If you add A:\ in resource access -> files, it changes to A:\*. In Sandboxie.ini, it is just A:\. I guess it is how it should be as per https://github.com/sandboxie-plus/Sandboxie/issues/2039#issuecomment-1193083437

"File and Registry entries auto append a '*' wildcard when they don't contain any. Prepending '|' disables this behaviour." [***](https://github.com/sandboxie-plus/Sandboxie/blob/a3ca0d27d00cc9b3be82a415d36cdf11fda81a46/SandboxiePlus/SandMan/Windows/OptionsAccess.cpp#L412-L416)

Ex.
`OpenFilePath=|A:\`

**Reply by ghost** [reply]
It does not work regardless if I put `UseRuleSpecificity=y` or not.

`\Device\VolumeA\\test.jpg.crdownload` is closed. Also, explorer shows info $Recycle.Bin is corrupted

**Reply by offhub** [reply]
If the application is installed in a sandbox, you should use `OpenPipePath` instead of `OpenFilePath`.

**Reply by ghost** [reply]
It is not. All I have tried do is to download any file to root of A:\ via Edge

**Reply by offhub** [reply]
I tested it, I guess there is a bug with file renaming. @DavidXanatos

**Reply by offhub** [reply]
#### Sandboxed CMD:
```
echo>CreateFile.txt
ren CreateFile.txt RenameFile.txt
```

(Double backslash?) -> \Device\ImDisk0\\\\RenameFile.txt (FD) 00120116.01.00000000
```
\Device\ImDisk0\CreateFile.txt
\Device\ImDisk0\CreateFile.txt (FA) 00120196.05.00000060
\Device\ImDisk0\CreateFile.txt
\Device\ImDisk0\CreateFile.txt
\Device\ImDisk0\CreateFile.txt
\Device\ImDisk0\
\Device\ImDisk0\ (FA) 00100001.01.00000021
\Device\ImDisk0\
\Device\ImDisk0\CreateFile.txt
\Device\ImDisk0\CreateFile.txt (FA) 00110080.01.00200020
\Device\ImDisk0\CreateFile.txt
\Device\ImDisk0\CreateFile.txt (FA) 00130116.01.00000020
\Device\ImDisk0\RenameFile.txt (FA) 00100080.01.00000021
\Device\ImDisk0\
\Device\ImDisk0\ (FA) 00120116.01.00000021
\Device\ImDisk0\RenameFile.txt (FA) 00100002.01.00000000
\Device\ImDisk0\\RenameFile.txt (FD) 00120116.01.00000000
\Device\ImDisk0\
```

**Reply by ghost** [reply]
Good to know this double backslash occurs to others as well. There was just one line with it though.

Is there any other way to achieve original request? Any workaround for this workaround?

**Reply by ghost** [reply]
I tested this with `WriteFilePath=A:\*\*` in `[GlobalSettings]`

and
```
UseRuleSpecificity=y
OpenFilePath=A:\*
```

in Sandbox settings. Result was the same

```
|Process|      |Type|     |Status|   |Value|                                        

20:39:58.567   File (K)   Closed     \Device\VolumeA\\OIP.jpg.crdownload   
```

**Reply by ghost** [reply]
I have noticed I don't need to Close/WriteFilePath anywhere and can just use `OpenFilePath=A:\abc`. Access to A:\ is closed completely (by default and for whatever reason) but abc folder.

**Comment by ghost** [comment]
OpenFilePath=Z:\a
ClosedFilePath=Z:\

I was unable to access Z:\a, however, I was able to save txt file in Z:\ and Z:\ subfolders. Those files have been saved outside of the box

**Reply by DavidXanatos** [reply]
did the files start with the letter a?

**Reply by ghost** [reply]
No. There is a glitch inside this glitch. To be honest I have no idea why folder 'a' is so special. I thought it is because it was created by user SID which does not exist on this PC and who was an admin.

icacls
```
a BUILTIN\Administrators:(I)(F)
  BUILTIN\Administrators:(I)(OI)(CI)(IO)(F)
  NT AUTHORITY\SYSTEM:(I)(F)
  NT AUTHORITY\SYSTEM:(I)(OI)(CI)(IO)(F)
  NT AUTHORITY\Authenticated Users:(I)(M)
  NT AUTHORITY\Authenticated Users:(I)(OI)(CI)(IO)(M)
  BUILTIN\Users:(I)(RX)
  BUILTIN\Users:(I)(OI)(CI)(IO)(GR,GE)
```

dir /q
`01.02.2023  20:45    <DIR>          ...                    a`

For now am not able to create one more glitched folder like this, so I will try something else

**Reply by DavidXanatos** [reply]
I cant reproduce the saving to Z: with that preset, was specificity on or of? did the files start with an a?

**Reply by ghost** [reply]
I will test it again as there was some issue with sandman not detecting changes in Sandboxie.ini, not reloading it and eventually I had to restart SBIE service

**Comment by ghost** [comment]
Experiment with
```
OpenFilePath=Z:\
ClosedFilePath=Z:\\*\\
```

21:45:07.644         File (D)   **Closed**     \Device\SomeVolumeZ\a
21:45:11.874         File (U)   Open       \Device\SomeVolumeZ\b
21:45:15.669         File (U)   Open       \Device\SomeVolumeZ\broken
21:45:16.593         File (U)   Open       \Device\SomeVolumeZ\broken\4 

If I manage to create one more folder like 'a', I may be able to achieve what I want... not exactly the way it was supposed to be I guess

**Comment by ghost** [comment]
@DavidXanatos Is sbie version with this fix available on Patreon?

**Reply by DavidXanatos** [reply]
its available on github: https://github.com/sandboxie-plus/Sandboxie/releases/tag/v1.8.0
